cmd/monitor/monitor.go

```diff
@@ -465,16 +440,29 @@ func renderMonitorUI(ctx context.Context, ec *ethclient.Client, ms *monitorStatu
 			return
 		}
 
+		if blockTable.SelectedRow == 0 {
+			ms.TopDisplayedBlock = ms.HeadBlock
+		}
+
 		if blockTable.SelectedRow == 0 || len(force) > 0 && force[0] {
 			allBlocks = updateAllBlocks(ms)
 			sort.Sort(allBlocks)
 		}
-		start := len(allBlocks) - windowSize - windowOffset
-		if start < 0 {
-			start = 0
+		toBlockNumber := ms.TopDisplayedBlock
+		fromBlockNumber := new(big.Int).Sub(toBlockNumber, big.NewInt(int64(windowSize-1)))
+		if fromBlockNumber.Cmp(zero) < 0 {
+			fromBlockNumber.SetInt64(0) // We cannot have block numbers less than 0.
+		}
+		renderedBlocksTemp := make([]rpctypes.PolyBlock, 0, windowSize)
+		for i := new(big.Int).Set(fromBlockNumber); i.Cmp(toBlockNumber) <= 0; i.Add(i, big.NewInt(1)) {
+			if block, ok := ms.BlockCache.Get(i.String()); ok {
+				renderedBlocksTemp = append(renderedBlocksTemp, block.(rpctypes.PolyBlock))
+			} else {
+				// If for some reason the block is not in the cache after fetching, handle this case.
+				log.Warn().Str("blockNumber", i.String()).Msg("Block should be in cache but is not")
```

Comment:
good catch. the blocks should already be there if the previous logic has worked out. this function ideally only renders the screen, and doesn't deal with actual block fetching. so far, i've been hitting this condition when debugging. i don't think we need a extra failsafe of fetching missing data here since the function name itself is `renderMonitorUI` !!